File: src/norms.c
/* $ID: matrix.c, last updated 2020-08-08, F.Osorio */

#include "base.h"
#include "norms.h"

void
norm_one(double *x, int *inc, int *n, double *value)
{ /* absolute-value norm */
  *value = F77_CALL(dasum)(n, x, inc);
}

void
norm_two(double *x, int *inc, int *n, double *value)
{ /* Euclidean norm */
  *value = F77_CALL(dnrm2)(n, x, inc);
}

void
norm_inf(double *x, int *inc, int *n, double *value)
{ /* infinity norm */
  int index;

  index  = F77_CALL(idamax)(n, x, inc);
  *value = fabs(x[index--]);
}

void
norm_minkowski(double *x, int *inc, int *n, double *p, double *value)
{ /* Minkowski norm */
  *value = F77_CALL(minkowski)(n, x, inc, p);
}

File: R/norms.R
## ID: norms.R, last updated 2020-08-07, F.Osorio

matrix.norm <- function(x, type = "Frobenius")
{ ## Computes a matrix norm
  if (is.data.frame(x))
    x <- as.matrix(x)
  if (!is.matrix(x))
    stop("supply a matrix-like 'x'")
  if (!is.numeric(x))
    stop("argument x is not a numeric matrix" )

  storage.mode(x) <- "double"
  dx <- dim(x)

  switch(type,
         "1" = {
           z <- .Fortran("maxcol_norm",
                         x = x,
                         ldx = as.integer(dx[1]),
                         n = as.integer(dx[1]),
                         p = as.integer(dx[2]),
                         val = as.double(0))$val
          },
          "Frobenius" = {
            n <- prod(dx)
            z <- .C("norm_two",
                    x = x,
                    inc = as.integer(1),
                    n = as.integer(n),
                    val = as.double(0))$val
          },
          "inf" = {
            z <- .Fortran("maxrow_norm",
                          x = x,
                          ldx = as.integer(dx[1]),
                          n = as.integer(dx[1]),
                          p = as.integer(dx[2]),
                          val = as.double(0))$val
          })
  z
}
